### ISSUE TITLE
CA2000: Rule does not trigger for IDisposable objects returned from factory methods

### ISSUE THREAD

**Issue by ozziepeeps** [dotnet/roslyn-analyzers#7590]
### Analyzer
Diagnostic ID: [CA2000](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000): Dispose objects before losing scope

### Analyzer source
SDK: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

Version: [SDK 8.0.307](https://dotnet.microsoft.com/en-us/download/dotnet/8.0)

### Describe the bug

The CA2000 rule currently does not trigger for objects implementing IDisposable when they are returned from a method, such as a factory method. This behavior is inconsistent, as the rule correctly triggers for IDisposable objects that are instantiated locally using a constructor. This can lead to potential resource leaks, as these objects may not be disposed of correctly.

### Steps To Reproduce

1) I have provided a repro in a GitHub repo here: https://github.com/ozziepeeps/DisposableIssue
2) Open the solution, build, and observe that only two instances of CA2000 are triggered. There should be four, per the comments

### Expected behavior
CA2000 is triggered for all for cases

### Actual behavior
CA2000 is triggered for two out of the four cases - the two cases that are local to the type

**Comment by ozziepeeps** [comment]
This is likely related to https://github.com/dotnet/roslyn-analyzers/issues/7487, which discusses the issue with `IAsyncDisposable` vs this issue that is for `IDisposable`

**Comment by sliekens** [comment]
I was wondering why CA2000 was not triggering for TPL Dataflow block links created with `LinkTo`, e.g.:

``` csharp
public static T ReceiveFromAny<T>(params ISourceBlock<T>[] sources)
{
    ArgumentNullException.ThrowIfNull(sources);
    WriteOnceBlock<T> writeOnceBlock = new(e => e);
    foreach (ISourceBlock<T> source in sources)
    {
        IDisposable link = source.LinkTo(writeOnceBlock, new DataflowLinkOptions { MaxMessages = 1 });
    }

    return writeOnceBlock.Receive();
}
```

I guess I will have a lot of CA2000 to fix if this issue is ever resolved
